integrations/optimum/src/haystack_integrations/components/embedders/optimum_document_embedder.py

```diff
@@ -0,0 +1,269 @@
+        deserialize_hf_model_kwargs(data["init_parameters"]["model_kwargs"])
+        return default_from_dict(cls, data)
+
+    def mean_pooling(self, model_output: torch.tensor, attention_mask: torch.tensor) -> torch.tensor:
+        """
+        Perform Mean Pooling on the output of the Embedding model.
+
+        :param model_output: The output of the embedding model.
+        :param attention_mask: The attention mask of the tokenized text.
+        :return: The embeddings of the text after mean pooling.
+        """
+        # First element of model_output contains all token embeddings
+        token_embeddings = model_output[0]
+        input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
+        sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)
+        sum_mask = torch.clamp(input_mask_expanded.sum(1), min=1e-9)
+        return sum_embeddings / sum_mask
+
+    def _prepare_texts_to_embed(self, documents: List[Document]) -> List[str]:
+        """
+        Prepare the texts to embed by concatenating the Document text with the metadata fields to embed.
+        """
+        texts_to_embed = []
+        for doc in documents:
+            meta_values_to_embed = [
+                str(doc.meta[key]) for key in self.meta_fields_to_embed if key in doc.meta and doc.meta[key] is not None
+            ]
+
+            text_to_embed = (
+                self.prefix + self.embedding_separator.join([*meta_values_to_embed, doc.content or ""]) + self.suffix
+            )
+
+            texts_to_embed.append(text_to_embed)
+        return texts_to_embed
+
+    def _embed_batch(self, texts_to_embed: List[str], batch_size: int) -> List[List[float]]:
+        """
+        Embed a list of texts in batches.
+        """
+        # Determine device for tokenizer output
+        device = (
+            "cuda"
+            if self.onnx_execution_provider
+            in ["CUDAExecutionProvider", "ROCMExecutionProvider", "TensorrtExecutionProvider"]
+            else "cpu"
+        )
+
+        # Sorting by length
+        length_sorted_idx = np.argsort([-len(sen) for sen in texts_to_embed])
+        sentences_sorted = [texts_to_embed[idx] for idx in length_sorted_idx]
+
+        all_embeddings = []
+        for i in tqdm(
+            range(0, len(sentences_sorted), batch_size), disable=not self.progress_bar, desc="Calculating embeddings"
+        ):
+            batch = sentences_sorted[i : i + batch_size]
+            encoded_input = self.tokenizer(batch, padding=True, truncation=True, return_tensors="pt").to(device)  # type: ignore
+
+            # Compute token embeddings
+            with torch.no_grad():
+                model_output = self.embedding_model(**encoded_input)  # type: ignore
+
+            # Perform mean pooling
+            sentence_embeddings = self.mean_pooling(model_output, encoded_input["attention_mask"].cpu())
```

Comment:
Why is the attention mask being moved to the host here? Shouldn't it be moved to `device` instead?

Reply:
You're right, we can put it directly on the device here. I will make that change.